jcommune-service/src/main/java/org/jtalks/jcommune/service/security/AclGroupPermissionEvaluator.java

```diff
@@ -113,33 +116,30 @@ public boolean hasPermission(Authentication authentication, Object targetId, Obj
     @Override
     public boolean hasPermission(Authentication authentication, Serializable targetId,
                                  String targetType, Object permission) {
-        boolean result = false;
         Long id = parseTargetId(targetId);
-
+        if (permission == ProfilePermission.EDIT_OWN_PROFILE &&
+                ((JCUser) authentication.getPrincipal()).getId() != id) {
+            return false;
+        }
+        JtalksPermission jtalksPermission;
         ObjectIdentity objectIdentity = aclUtil.createIdentity(id, targetType);
-        Permission jtalksPermission;
+        Sid sid = sidFactory.createPrincipal(authentication);
+        List<AccessControlEntry> aces;
+        List<GroupAce> controlEntries;
         if (permission instanceof Permission) {
-            jtalksPermission = (Permission) permission;
+            jtalksPermission = (JtalksPermission) permission;
         } else {
             jtalksPermission = getPermission(permission);
         }
-        Sid sid = sidFactory.createPrincipal(authentication);
-        List<AccessControlEntry> aces;
-        List<GroupAce> controlEntries;
 
         try {
             aces = ExtendedMutableAcl.castAndCreate(mutableAclService.readAclById(objectIdentity)).getEntries();
-            controlEntries = aclManager.getGroupPermissionsOn(objectIdentity);
+            controlEntries = aclManager.getGroupPermissionsFilteredByPermissionOn(objectIdentity, jtalksPermission);
         } catch (NotFoundException nfe) {
             aces = new ArrayList<>();
             controlEntries = new ArrayList<>();
         }
 
-        if (permission == ProfilePermission.EDIT_OWN_PROFILE &&
-                ((JCUser) authentication.getPrincipal()).getId() != id) {
-            return false;
-        }
-
         if (isRestrictedForSid(sid, aces, jtalksPermission) ||
                 isRestrictedForGroup(controlEntries, authentication, jtalksPermission) ||
                 isRestrictedPersonalPermission(authentication, jtalksPermission)) {
```

Comment:
I thought we were going to check if the permission is a personal permission before checking the rest.. 

Though I see a check below. But I think that if it _is_ a PersonalPermission, then it doesn't make sense to check others. So it makes sense to put that check here.